pkg/ebpf/uprobes/attacher.go

```diff
@@ -0,0 +1,721 @@
+	thisPID, err := kernel.RootNSPID()
+	if err != nil {
+		return fmt.Errorf("error getting PID of our own process: %w", err)
+	}
+
+	err = kernel.WithAllProcs(ua.config.ProcRoot, func(pid int) error {
+		if pid == thisPID { // don't scan ourselves
+			return nil
+		}
+
+		return ua.AttachPIDWithOptions(uint32(pid), true)
+	})
+
+	return err
+}
+
+// handleProcessStart is called when a new process is started, wraps AttachPIDWithOptions but ignoring the error
+// for API compatibility with processMonitor
+func (ua *UprobeAttacher) handleProcessStart(pid uint32) {
+	_ = ua.AttachPIDWithOptions(pid, false) // Do not try to attach to libraries on process start, it hasn't loaded them yet
+}
+
+// handleProcessExit is called when a process finishes, wraps DetachPID but ignoring the error
+// for API compatibility with processMonitor
+func (ua *UprobeAttacher) handleProcessExit(pid uint32) {
+	_ = ua.DetachPID(pid)
+}
+
+func (ua *UprobeAttacher) handleLibraryOpen(event *ebpf.DataEvent) error {
+	defer event.Done()
+
+	libpath := sharedlibraries.ToLibPath(event.Data)
+	path := sharedlibraries.ToBytes(&libpath)
+
+	err := ua.AttachLibrary(string(path), libpath.Pid)
+
+	return err
+}
+
+// AttachLibrary attaches the probes to the given library, opened by a given PID
+func (ua *UprobeAttacher) AttachLibrary(path string, pid uint32) error {
+	if (ua.config.ExcludeTargets&ExcludeSelf) != 0 && int(pid) == os.Getpid() {
+		return ErrSelfExcluded
+	}
+
+	matchingRules := ua.getRulesForLibrary(path)
+	if len(matchingRules) == 0 {
+		return ErrNoMatchingRule
+	}
+
+	registerCB := func(path utils.FilePath) error {
+		return ua.attachToBinary(path, matchingRules)
+	}
+	unregisterCB := func(path utils.FilePath) error {
+		return ua.detachFromBinary(path)
+	}
+
+	return ua.fileRegistry.Register(path, pid, registerCB, unregisterCB)
+}
+
+// getRulesForLibrary returns the rules that match the given library path
+func (ua *UprobeAttacher) getRulesForLibrary(path string) []*AttachRule {
+	var matchedRules []*AttachRule
+
+	for _, rule := range ua.config.Rules {
+		if rule.matchesLibrary(path) {
+			matchedRules = append(matchedRules, rule)
+		}
+	}
+	return matchedRules
+}
+
+// getRulesForExecutable returns the rules that match the given executable
+func (ua *UprobeAttacher) getRulesForExecutable(path string) []*AttachRule {
+	var matchedRules []*AttachRule
+
+	for _, rule := range ua.config.Rules {
+		if rule.matchesExecutable(path) {
+			matchedRules = append(matchedRules, rule)
+		}
+	}
+	return matchedRules
+}
+
+// getExecutablePath resolves the executable of the given PID looking in procfs. Automatically
+// handles delays in procfs updates. Will return an error if the path cannot be resolved
+func (ua *UprobeAttacher) getExecutablePath(pid uint32) (string, error) {
+	pidAsStr := strconv.FormatUint(uint64(pid), 10)
+	exePath := filepath.Join(ua.config.ProcRoot, pidAsStr, "exe")
+
+	var binPath string
+	err := errors.New("iteration start")
```

Comment:
Currently you're regenerating and allocating the default error
it should be defined only once, outside of a function scope